### ISSUE TITLE
3.2 Release Plan

### ISSUE THREAD

**Issue by sksamuel** [kotlintest/kotlintest#482]
[ ] Kotlin 1.3
[ ] Coroutine Support
[ ] Isolation Levels #379 	
[ ] Improved shrinker support
[ ] Co-variant Gens #471
[ ] Package selectors in discovery #461 
[ ] Upgrade from reflections to classgraph
[ ] output to show the test cased generated in Property testing or table testing
[ ] Better support for comparing multi-line strings #402 
[ ] Arrow to 0.80
[ ] Remember previous failing test cases #388
[ ] New matchers for 3.2 #325
[ ] Improved listener support for disabled tests #468 #458

**Comment by sksamuel** [comment]
Please add anything you think is missing @Kerooker

**Comment by Kerooker** [comment]
Agreed. Android Integration tests can come later.

But if we're close to releasing 3.2, maybe there's no need for a 3.1.11?

**Comment by sksamuel** [comment]
Well yeah not really, but let's do it anyway in case some people don't want to "worry" about upgrading to a new version and just want a bug release. I'm happy to do the cherry-picking if you want.

**Comment by Kerooker** [comment]
I'm not very sure on how to do it, never done it before :P

I'm a little bit confused on how to keep the authors and the commits, while still doing the merge correctly.

**Comment by sksamuel** [comment]
You just need to find the actual commit sha, not the merge commit, and do `git cherry-pick sha`. I'll pick it up, won't take long, although it's 3am here so I might wait.

**Comment by sksamuel** [comment]
We can probably bump the shrinker support too.

**Comment by Kerooker** [comment]
Agreed

**Comment by Kerooker** [comment]
Now that https://github.com/kotlintest/kotlintest/issues/489 was release, what's next?

Do you think we need to drop any of the remaining issues and deliver 3.2 faster?

**Comment by sksamuel** [comment]
Probably just coroutines.

**Comment by dave08** [comment]
Quite a few people are using coroutines now, and it's kinda hard to just make an extension function for it since some of the things needed are `protected`... will support for it come out soon?

**Comment by Kerooker** [comment]
@dave08 
Could you provide me with a simple test case for your coroutines usage? I'm trying to design a test case for them, but I'm afraid to create a scenario that's not close to the real one.

I'm not very used to coroutines. What would I need to do? Run any `suspend` function from inside a test? So the test case

```
"Testing coroutines" {
    val suspendReturn = mySuspendFunction()

   suspendReturn shouldBe "Foo"
}
```

would kinda work?

**Comment by dave08** [comment]
@Kerooker The most basic way to test in coroutines needs to surround each test with `runBlocking { }` which also provides a `CoroutineScope` to `launch` or `async` (and cancels all those automatically using structured concurrency) from and also gives the ability to run suspend functions. In your case that the reciever is already taken, you could perhaps provide the coroutinescope from runblocking as the `it` of the test.

**Comment by Kerooker** [comment]
And if `TestContext` is a `CoroutineScope`, I suppose it won't be necessary to add that `scope ->`, correct?

**Comment by dave08** [comment]
Correct. You have a prototype out I could see (like a branch or fork commit)?

I'm not sure if it's a good idea to use runBlocking even on regular tests, maybe a special infix like co (like in mockk...) to add it if a tests needs it.

Also in 1.0.0-alpha coroutines I think they added a test module, and the plans are to support a special TestDispatcher like the test scheduler in Rx, so it might be nice to keep a door open for such things

**Comment by sksamuel** [comment]
I started some work to make each `TestContext` a scope, but it gets complicated which is why it's parked atm.

We need it to work not only for suspend functions but also allowing async in there.

**Comment by Kerooker** [comment]
I'm playing with your branch, @sksamuel

**Comment by Kerooker** [comment]
@dave08 There's a specific branch called feature/coroutines in which we're trying some things

**Comment by sksamuel** [comment]
Ok cool, but I parked it because I think it's gonna need a considerably more work.
Notably, I think the execution of the tests themselves can be done as a coroutine rather than in an executor. Then the coroutine context is propagated all the way down to the user's scope.

**Comment by Kerooker** [comment]
That's what I was thinking too

**Comment by dave08** [comment]
In Spek, I could do:
```
fun <T> SpecBody.coMemoized(
		mode: CachingMode = CachingMode.TEST,
		factory: suspend () -> T
): LifecycleAware<T> = memoized(mode) { runBlocking { factory() } }

fun TestContainer.coIt(description: String, body: suspend TestBody.(scope: CoroutineScope) -> Unit) =
	test("it $description", body = { runBlocking { body(this) } })
```
Would there be a way to do this in KotlinTest?

**Comment by dave08** [comment]
A *very* simple way to introduce support would be to make a file with a bunch of extension methods like this:
```
fun AbstractDescribeSpec.DescribeScope.coIt(name: String, test: suspend TestContext.(CoroutineScope) -> Unit) =
        it(name) { runBlocking { test(this) }}
```
It might not work for all test types though...

**Comment by Kerooker** [comment]
Why is it better than just do `runBlocking` in everything? We wouldn't need
to do that

**Comment by dave08** [comment]
If the core runs on Coroutines, then you're right. But if not, why have the extra overhead when not all tests will be using Coroutines?

**Comment by Kerooker** [comment]
It's not that much overhead to run all in a blocked coroutine. It would
prevent users to have to add an extra keyword in the code. Do you think any
complication would happen due to this?

**Comment by dave08** [comment]
You need real threads for when users call blocking api, the main advantage of coroutines is when using async api that can suspend and free up the thread it's runnning on for something else to use it. There's also a big distinction between cpu intensive computations that need `Dispatchers.Default` (# of cpus - 1 thread pool) and network/disk operations that when suspendable, can use `Dispatchers.IO` (growing thread pool according to needs), how will the user provide this, and a user that isn't familiar with coroutines will be lost...

**Comment by dave08** [comment]
Also, another big point to consider is cancellation in structured concurrency, since an exception might end up cancelling its outer scope, and all the child coroutines (even from the other tests running in parallel). I'm not sure if that's the case with runBlocking in your case, but if so, you would need the root scope to have a SupervisorJob and have sub scopes surrounded by the `coroutineScope { }` builder...

**Comment by sksamuel** [comment]
@dave08 take a look at my implementation in #518 

I would appreciate your feedback to see if I've left any holes.

Each spec is executed in it's own thread, and the actual test case is delegated out to a coroutine context created from a separate thread pool, so users writing blocking code won't result in a deadlock in the test framework itself.

The DSL allows suspendable functions everywhere you'd expect.

**Comment by sksamuel** [comment]
If we're happy with #518 after review + feedback changes, we can include in 3.2. We could drop #411 and I can finish #494 and then release 3.2.RC1 ?

**Comment by dave08** [comment]
@sksamuel I'm a bit busy to go through this very thoroughly now, but I think you need to use a SupervisorScope around the whole test execution, and just surround each test with `coroutineScope { }` for the cancellation problem, if launch is done within a scope I don't think you need to join it. I might find more things if I take a better look, and I don't know junitplatform well enough to say that these comments are really relevant... I'll see what I can do a bit later on...

**Comment by sksamuel** [comment]
The problem with that is that coroutines use co-operative cancellation, so job.cancel won't cut the mustard.

**Comment by dave08** [comment]
@sksamuel What do you mean? If you use `coroutineScope { }` around each test, and the main scope has a `SupervisorJob`, the main scope won't be cancelled if the test scope was, but if you cancel the main scope (or its SupervisorJob), it'll cancel all of its children. You need some other behavior than this?

**Comment by sksamuel** [comment]
What I mean is this.

```
fun main() {
  runBlocking {
    val job = launch {
      Thread.sleep(200000)
    }
    Executors.newSingleThreadExecutor().submit {
      Thread.sleep(2000)
      job.cancel()
    }
    job.join()
    println("Done")
  }
}
```

`job.cancel` is not a drop in replacement for Thread.interrupt(). It's meant to be used with co-operative cancellation, and that isn't good enough for our requirements. I can't expect users to pepper their code with yield() etc, and most code that blocks won't be a coroutine anyway.

**Comment by dave08** [comment]
Got you now, so how did you currently solve that problem @sksamuel ?

**Comment by dave08** [comment]
In `InstancePerLeafSpecRunner` you use the `runBlocking` scope (`this`), but in `SingleInstanceSpecRunner` you use `GlobalScope` any reason for it @sksamuel ? `GlobalScope` is usually used for daemon type coroutines that execute the whole lifetime of the app...

**Comment by sksamuel** [comment]
That was a leftover, I've changed it to use this.

In that link, it's using a dispatcher created via `newSingleThreadContext(name)` which is essentially what I do but via a single thread executor as an executor is nicer to use than a basic thread.

**Comment by Kerooker** [comment]
I think  it's all ready for a snapshot release?

**Comment by sksamuel** [comment]
RC1 because we're not going to add / remove functionality, just fix bugs, and if none are found, it'll be final.
Which is what a release candidate is.

**Comment by Kerooker** [comment]
Uh... I accidentally pressed something, and I think it creates a release (was trying to configure tag to display as pre-release).

Is that ok? x.x
